iroh-net/src/magicsock.rs

```diff
@@ -993,24 +994,30 @@ impl MagicSock {
         }
     }
 
-    fn poll_send_ping(&self, ping: &SendPing, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
-        let SendPing {
-            id,
-            dst,
-            dst_node,
-            tx_id,
-            purpose,
-        } = ping;
-        let msg = disco::Message::Ping(disco::Ping {
-            tx_id: *tx_id,
-            node_key: self.public_key(),
-        });
-        ready!(self.poll_send_disco_message(dst.clone(), *dst_node, msg, cx))?;
-        let msg_sender = self.actor_sender.clone();
-        debug!(%dst, tx = %hex::encode(tx_id), ?purpose, "ping sent (polled)");
-        self.node_map
-            .notify_ping_sent(*id, dst.clone(), *tx_id, *purpose, msg_sender);
-        Poll::Ready(Ok(()))
+    /// Tries to send the ping actions.
+    ///
+    /// Note that on failure the (remaining) ping actions are simply dropped.  That's bad!
+    /// The Endpoint will think a full ping was done and not request a new full-ping for a
+    /// while.  We should probably be buffering the pings.
+    fn try_send_ping_actions(&self, msgs: Vec<PingAction>) -> io::Result<()> {
+        for msg in msgs {
+            // Abort sending as soon as we know we are shutting down.
+            if self.is_closing() || self.is_closed() {
+                return Ok(());
+            }
```

Comment:
Is this here to prevent cases where we `try_send_ping` even though `self.is_closing()`?
Are these cases bad?
And if this is here to prevent that, then I don't think it's doing that, as the endpoint closing & this running could still race.

Is this perhaps meant to be outside the `for` loop?
Or can `try_send_ping` cause `self.is_closing()`?

In any case I'm a little confused by this.

Reply:
this is here because... history?  maybe from the go code still?  Maybe it was a bug fix early on?  I actually suspect the last one.

I don't care too much about this kind of check, it appears in a few places.  Ultimately I think the architecture should be robust enough to deal without this.  But as long as we have a "closed" bool in the state we'll have this kind of thing.  I think it's fine to leave in.

Reply:
Okay, let's leave it :)